Update for Neo4J 3.x, neo4jrb 8.x, and neojrb-core 7.0.x

TyGuy opened the pull request:
This is a PR that I started out of need for this support in a personal project. It is certainly pretty rough but I think a good start, and I'm willing to clean it up to get this working. It replaces the old providers with `spatial.procedure` calls.

Could be a fix for: https://github.com/neo4jrb/neo4jrb_spatial/issues/17

### Notes/Questions:
* I totally changed the api for this, since it seems like the underlying neo4j_spatial api has changed pretty significantly (for example, adding nodes to layers, as opposed to adding nodes to indexes). I want to know what the maintainers think; I could definitely see the benefit of an in-between patch, where we do the "new way" under the hood, but maintain most of the old method names, and provide a deprecation warning. But I feel like eventually getting rid of those names (i.e. `add_node_to_index`, `create_index`, etc.) would make sense. What are people's thoughts here?
* Should there be a callback in ActiveNode that adds the node to the layer as an `after_create` hook? I saw a commented-out attempt to do this but never got it working.
* I also exposed and tweaked the `install` rake task, to try and make it a little more easily useable for installing neo4j_spatial (could move that into another PR if people prefer)
* What else do I need to do to get this to be mergeable?

### Todos:
* [x] make the ActiveNode portion of the gem use scopes instead of methods, so we don't have to prepend with `all` (as in `Model.all.<thing>`)
* [x] update README to include explanations/basic usage of new methods

This is a little rough but please provide feedback and we can take it from there! Thanks.

Ty

TyGuy committed:
Update for newer version of neo4jrb and neo4jrb-core gems.

+ use spatial procedures instead of old REST posts
+ update gemfile for newer versions

TyGuy committed:
Make rake task available outside gem

TyGuy committed:
Make versions more flexible

TyGuy committed:
Make version failure more helpful

TyGuy committed:
Default to neo4j installed version so NEO4J_VERSION is not required

TyGuy committed:
Match new procedure calls

TyGuy committed:
Add ActiveNode methods and tests

TyGuy committed:
update rubocop and clean unused code

cheerfulstoic commented:
Having a look at this now, though I've not used this project yet, but I can definitely give feedback on the code.  Happy to merge / release once it looks good (I generally like the idea of the [C4.1](https://rfc.zeromq.org/spec:22/C4/) community process)

cheerfulstoic commented on lib/neo4jrb_spatial/rake_tasks/neo4j_spatial.rake:
```diff
@@ -0,0 +1,68 @@
+  versions.map! { |e| e.first.split('-neo4j-') }
+  versions.select { |e| match_version?(version, e.last) }.last
+
+  compatible_versions = versions.select { |e| match_version?(version, e.last) }.last
```
Should be `compatible_version` since it's not an array

cheerfulstoic commented on lib/neo4jrb_spatial/rake_tasks/neo4j_spatial.rake:
```diff
@@ -0,0 +1,68 @@
+  versions = Net::HTTP.get_response(URI.parse(uri)).body
+  versions = versions.scan(/<version>([a-z\-0-9\.]+)<\/version>/)
+  versions.map! { |e| e.first.split('-neo4j-') }
+  versions.select { |e| match_version?(version, e.last) }.last
```
This seems like this wouldn't actually accomplish anything because it's not being assigned.  Is this left over from the `compatible_versions = ` below?  If so we wouldn't need to extend the size of the `Metrics/AbcSize` rubocop rule

cheerfulstoic commented on .rubocop.yml:
```diff
@@ -21,6 +21,9 @@ Style/SpaceInsideHashLiteralBraces:
 
+Metrics/AbcSize:
+  Max: 20
+
```
I always really, really try to avoid allowing the code to get more complex on a global level.  Rarely I'll make a `rubocop:disable` comment around the method itself if I absolutely need to, but most of the time I can find a way to refactor and most of the time my code is better for it (not always ;) )

cheerfulstoic replied:
Also, I'd suggest running `rubocop -aD`.  That will fix any issues that it can, and then you can resolve the rest.  The `D` will output the cop's name so that you can look it up in case you don't know what it's talking about ;)

cheerfulstoic commented on .travis.yml:
```diff
@@ -8,21 +8,21 @@ install:
   - "bundle exec rake neo4j_spatial:install"
   - "bundle exec rake neo4j:start --trace"
   - "sleep 20"
+  - "bundle exec rake neo4j:generate_schema_migration[constraint,Restaurant,uuid]"
```
You can certainly do this, but the specs won't work on other peoples' machines unless they know to run this line.    I generally try to do something like this inside the specs themselves:

```ruby
Neo4j::ActiveBase.label_object(: Restaurant).create_constraint(:uuid, type: :unique)
```

You might also rescue from `Neo4j::Core::CypherSession::SchemaErrors::IndexAlreadyExistsError` so that if it already exists it doesn't just fail

cheerfulstoic commented on .travis.yml:
```diff
@@ -8,21 +8,21 @@ install:
   global:
     - JRUBY_OPTS="-J-Xmx1280m -Xcompile.invokedynamic=false -J-XX:+TieredCompilation -J-XX:TieredStopAtLevel=1 -J-noverify -Xcompile.mode=OFF"
-    - NEO4J_VERSION=3.0.1
+    - NEO4J_VERSION=3.1.0
```
3.2 just came out, BTW ;)

cheerfulstoic commented on README.md:
````diff
@@ -136,3 +142,29 @@ mostly works for an idea of the basics, just replace Neography-specific commands
+...
+```
+
+To resolve, simply run the commands as it says (but maybe prefix `bundle exec` depending on your setup):
````
Ah, I see that you have this in the docs (which is great!).  It would be nicer, though, if people didn't need to think about it

cheerfulstoic commented on lib/neo4j/active_node/spatial.rb:
```diff
@@ -5,38 +5,75 @@ def self.included(other)
+
+          @spatial_layer_name = layer_name
+          @spatial_layer_type = options.fetch(:type, 'SimplePoint')
+          @spatial_layer_config = options.fetch(:config, 'lon:lat')
```
What about having a `Struct` instead?  Something like:

```ruby
attr_reader :spatial_layer

SpatialLayer = Struct.new(:name, :type, :config)
def spatial_layer(layer_name = nil, options = {})
  return spatial_layer_name unless layer_name

  @spatial_layer = SpatialLayer.new(layer_name, options.fetch(:type, 'SimplePoint'), options.fetch(:config, 'lon:lat'))
end

cheerfulstoic commented on lib/neo4j/active_node/spatial.rb:
```diff
@@ -5,38 +5,75 @@ def self.included(other)
-        # private :create_index_callback
+          lon_name, lat_name = spatial_layer_config.split(':')
+
+          Neo4j::ActiveBase.current_session.add_layer(spatial_layer_name, spatial_layer_type, lat_name, lon_name)
```
This might just be part of the original API, but it feels strange to extend the session objects with `add_layer` / `remove_layer` / `within_distance` / etc...  I would expect maybe some other object which is given a session on initialization.  Depends on if you find that worthwhile and if you want to go to the extent of changing it

cheerfulstoic commented on lib/neo4j/spatial.rb:
```diff
@@ -1,143 +1,180 @@
+  module Core
+    # rubocop:disable Metrics/ModuleLength
     module Spatial
+      # rubocop:enable Metrics/ModuleLength
```
I'm happier having this disabled locally than globally, but I still really don't like disabling rubocop rules

cheerfulstoic commented on lib/neo4j/spatial.rb:
```diff
@@ -1,143 +1,180 @@
+      def spatial_procedures
+        call_query = 'CALL spatial.procedures() YIELD name'
+
+        query(call_query, {}).map(&:name)
```
What about:

```ruby
query('CALL spatial.procedures() YIELD name').map(&:name)
```

cheerfulstoic commented on lib/neo4j/active_node/spatial.rb:
```diff
@@ -5,38 +5,75 @@ def self.included(other)
+
+          Neo4j::ActiveBase.current_session
+            .intersects(layer, geometry, execute: false)
+            .proxy_as(model, :node)
```
Again, I'm not familiar with this gem, but using `Neo4j::ActiveBase` and `proxy_as` makes the gem dependent on the `neo4j` gem and thus couldn't be used with the `neo4j-core` gem by itself.  That doesn't seem to be a big deal because `proxy_as` was used previously, but just thought I'd mention it

cheerfulstoic commented on lib/neo4j/spatial.rb:
```diff
@@ -1,143 +1,180 @@
+                    .where('id(n) = {node_id}')
+                    .with(:n).call('spatial.addNode({layer}, n) YIELD node')
+                    .return('node')
+                    .params(options)
```
I would call this variable `params` or `parameters` instead of `options` because `options` is generally for a hash which affects the behavior of a method.  Also, you could do:

```ruby
procedure = query_.match(:n)
            .where(n: {neo_id: node.neo_id})
            .with(:n).call('spatial.addNode({layer}, n) YIELD node').params(layer: layer)
            .return(:node)
```

cheerfulstoic commented on lib/neo4j/spatial.rb:
```diff
@@ -1,143 +1,180 @@
-        parse_response! Neo4j::Session.current.connection.post("/db/data/#{path}", options).body
+        query_ = Query.new(session: self)
+        query_.call(call_query).params(procedure_args)
       end
```
These two methods could probably be dried up, maybe with a third argument of `, with_node = true`

cheerfulstoic commented on lib/neo4j/spatial.rb:
```diff
@@ -1,143 +1,180 @@
+                      spatial_procedure(procedure_name, procedure_args)
+                    else
+                      spatial_procedure_without_node(procedure_name, procedure_args)
+                    end
```
If you added the third argument above, this would be a lot simpler

TyGuy commented:
@cheerfulstoic thank you for all the comments. This is super helpful. I am going to be traveling over the next few weeks, but I will try to address these all in the next week or so, if I can find some down time. Worst case I'll get to it in about 3 weeks

cheerfulstoic commented:
:+1: ;)

TyGuy committed:
Address PR comments

+ remove rubocop exceptions
+ make constraint creation automatic
+ couple smaller refactors

TyGuy committed:
Change ActiveNode methods to scopes

TyGuy committed:
Fix bugs

+ require active support,
+ fix ‘importShapefileToLayer’ call
+ improve test cleanup

TyGuy committed:
Update Readme

TyGuy commented:
@cheerfulstoic I have made changes. Let me know what you think about them, and what else could use improvement.

ALSO we should be thoughtful about how we merge/release this (I think a minor or major version bump is in order, but honestly I don't know much about versioning). Because this will break with any of (neo4jrb < 8), (neo4jrb-core < 7), (neo4j < 3). I have updated those requirements in the README, but people might be in for a bad surprise otherwise.

cheerfulstoic commented:
Yep, since there's a chance that it'll break in older versions, I'll make it a major version bump.  Thanks!  Going to release now

cheerfulstoic commented:
Released as 2.0.0!  Thanks again!

ProGM commented:
Thank you very much for your work on this @TyGuy ;)
